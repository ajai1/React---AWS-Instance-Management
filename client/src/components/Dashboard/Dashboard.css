.dashboard__container {
  height: 100%;
  background-color: #e9f0f0;
  display: flex;
  flex-direction: column;
}

.view__container {
  display: flex;
  width: 100%;
  height: 100%;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
}

.instance__container {
  margin-top: 20px;
  width: 80%;
  height: 100px;
  background-color: white;
  border: 0.5px solid rgba(0, 0, 0, 0.322);
  border-radius: 15px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 10px 20px;
}

.instance__table {
  width: 100%;
  margin-top: 15px;
}

.instance__detail__container {
  margin-top: 20px;
  width: 80%;
  background-color: white;
  border: 0.5px solid rgba(0, 0, 0, 0.322);
  border-radius: 15px;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  padding: 20px;
}

.typography__container {
  display: flex;
  flex-direction: row;
}

.switch__box {
  display: flex;
  align-items: center;
  margin-right: 20px;
}

@media only screen and (max-width: 600px) {
  .instance__container {
    height: auto;
    flex-direction: column;
  }
}
