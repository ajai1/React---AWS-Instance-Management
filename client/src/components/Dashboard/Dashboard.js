import React, { useState, useEffect } from "react";
import Header from "./Header/Header";
import "./Dashboard.css";
import { Switch, Typography } from "@material-ui/core";
import axios from "axios";

import TableComponent from "./TableComponent/TableComponent";

const Dashboard = () => {
  const [instances, setInstances] = useState([]);
  const [isUSD, setIsUSD] = useState(true);
  const [charges, setCharges] = useState([0.19, 0.35]);

  useEffect(() => {
    getInstances();
  }, []);

  useEffect(() => {
    const runningInstances = instances
      .filter((instance) => instance.status === "running")
      .reduce((a, b) => {
        return a + b.costPerHour;
      }, 0);

    const stoppedInstances = instances
      .filter((instance) => instance.status === "stopped")
      .reduce((a, b) => {
        return a + b.costPerHour;
      }, 0);

    setCharges([runningInstances.toFixed(3), stoppedInstances.toFixed(3)]);
  }, [instances]);

  const getInstances = async () => {
    const res = await axios.get("/api/instances", {
      headers: {
        Authorization: "Bearer " + localStorage.getItem("authorization"),
      },
    });
    setInstances(res.data.instances);
    return res.data;
  };

  const startInstance = async (id) => {
    const res = await axios.get(`/api/instances/start/${id}`, {
      headers: {
        Authorization: "Bearer " + localStorage.getItem("authorization"),
      },
    });
    getInstances();
    return res.data;
  };

  const stopInstance = async (id) => {
    const res = await axios.get(`/api/instances/stop/${id}`, {
      headers: {
        Authorization: "Bearer " + localStorage.getItem("authorization"),
      },
    });
    getInstances();
    return res.data;
  };

  const onCurrencyChange = () => {
    let newCharges;
    if (isUSD) {
      newCharges = charges.map((charge) => (charge * 72.96).toFixed(2));
    } else {
      newCharges = charges.map((charge) => (charge / 72.96).toFixed(2));
    }
    setCharges(newCharges);
    setIsUSD(!isUSD);
  };

  return (
    <div className="dashboard__container">
      <div>
        <Header />
      </div>
      <div className="view__container">
        {/* Instance Info Container */}
        <div className="instance__container">
          <div className="typography__container">
            <div style={{ marginLeft: "20px" }}>
              <Typography variant="h6">
                {isUSD ? "$" : "₹"}
                {charges[0]}/hr
              </Typography>
              <Typography variant="body1">Running Instances</Typography>
            </div>
            <div style={{ marginLeft: "20px" }}>
              <Typography variant="h6">
                {isUSD ? "$" : "₹"}
                {charges[1]}/hr
              </Typography>
              <Typography variant="body1">Stopped Instances</Typography>
            </div>
          </div>
          <div className="switch__box">
            <Typography variant="body1">INR</Typography>
            <Switch
              color="primary"
              checked={isUSD}
              onChange={onCurrencyChange}
            />
            <Typography variant="body1">USD</Typography>
          </div>
        </div>
        {/* Instance Table Container */}
        <TableComponent
          instances={instances}
          startInstance={startInstance}
          stopInstance={stopInstance}
          isUSD={isUSD}
        />
      </div>
    </div>
  );
};

export default Dashboard;
