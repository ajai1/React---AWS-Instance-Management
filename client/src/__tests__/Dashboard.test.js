import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import Dashboard from "../components/Dashboard/Dashboard";

it("Match the Layout", () => {
  const { asFragment } = render(<Dashboard />);
  expect(asFragment()).toMatchSnapshot();
});
