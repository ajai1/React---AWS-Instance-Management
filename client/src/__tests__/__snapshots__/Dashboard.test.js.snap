// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match the Layout 1`] = `
<DocumentFragment>
  <div
    class="dashboard__container"
  >
    <div>
      <div>
        <header
          class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
        >
          <div
            class="MuiToolbar-root MuiToolbar-regular header__container MuiToolbar-gutters"
          >
            <div>
              <h6
                class="MuiTypography-root MuiTypography-h6"
              >
                Dashboard
              </h6>
            </div>
            <div>
              <p
                class="MuiTypography-root MuiTypography-body1"
                data-testid="logout"
                style="cursor: pointer;"
              >
                Logout
              </p>
            </div>
          </div>
        </header>
      </div>
    </div>
    <div
      class="view__container"
    >
      <div
        class="instance__container"
      >
        <div
          class="typography__container"
        >
          <div
            style="margin-left: 20px;"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6"
            >
              $0.000/hr
            </h6>
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              Running Instances
            </p>
          </div>
          <div
            style="margin-left: 20px;"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6"
            >
              $0.000/hr
            </h6>
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              Stopped Instances
            </p>
          </div>
        </div>
        <div
          class="switch__box"
        >
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            INR
          </p>
          <span
            class="MuiSwitch-root"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-checked-2 Mui-checked"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  checked=""
                  class="PrivateSwitchBase-input-4 MuiSwitch-input"
                  type="checkbox"
                  value=""
                />
                <span
                  class="MuiSwitch-thumb"
                />
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiSwitch-track"
            />
          </span>
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            USD
          </p>
        </div>
      </div>
      <div
        class="instance__detail__container"
      >
        <div>
          <h6
            class="MuiTypography-root MuiTypography-h6"
          >
            Instances
          </h6>
        </div>
        <div
          class="instance__table"
        >
          <div
            class="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
          >
            <table
              aria-label="customized table"
              class="MuiTable-root"
            >
              <thead
                class="MuiTableHead-root"
              >
                <tr
                  class="MuiTableRow-root MuiTableRow-head"
                >
                  <th
                    class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-5"
                    scope="col"
                  >
                    ID
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-5 MuiTableCell-alignLeft"
                    scope="col"
                  >
                    Instance Name
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-5 MuiTableCell-alignRight"
                    scope="col"
                  >
                    Cost Per Hour
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-5 MuiTableCell-alignLeft"
                    scope="col"
                  >
                    Status
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-5 MuiTableCell-alignRight"
                    scope="col"
                  >
                    Action
                  </th>
                </tr>
              </thead>
              <tbody
                class="MuiTableBody-root"
              />
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
